probe/awsecs/client.go

```diff
@@ -31,72 +70,146 @@ func newClient(cluster string) (*ecsClient, error) {
+	now := time.Now()
+	deploymentIDs := make([]string, 0, len(service.Deployments))
+	for _, deployment := range service.Deployments {
+		deploymentIDs = append(deploymentIDs, *deployment.Id)
+	}
+	return ecsService{
+		serviceName:       *service.ServiceName,
+		deploymentIDs:     deploymentIDs,
+		desiredCount:      *service.DesiredCount,
+		pendingCount:      *service.PendingCount,
+		runningCount:      *service.RunningCount,
+		taskDefinitionARN: *service.TaskDefinition,
+		fetchedAt:         now,
+		lastUsedAt:        now,
 	}
+}
+
+// Returns a channel from which service ARNs can be read.
+// Cannot fail as it will attempt to deliver partial results, though that may end up being no results.
+func (c ecsClient) listServices() <-chan string {
+	log.Debugf("Listing ECS services")
+	results := make(chan string)
+	go func() {
+		count := 0
+		err := c.client.ListServicesPages(
+			&ecs.ListServicesInput{Cluster: &c.cluster},
+			func(page *ecs.ListServicesOutput, lastPage bool) bool {
+				for _, arn := range page.ServiceArns {
+					count++
+					results <- *arn
+				}
+				return true
+			},
+		)
+		if err != nil {
+			log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
+		}
+		log.Debugf("Listed %d services", count)
+		close(results)
+	}()
 	return results
 }
 
-// cannot fail as it will attempt to deliver partial results, though that may end up being no results
-func (c ecsClient) getServices() map[string]*ecs.Service {
-	results := map[string]*ecs.Service{}
-	lock := sync.Mutex{} // lock mediates access to results
+// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
+// with full ecsService objects being put into the cache. Closes done when finished.
+func (c ecsClient) describeServices() (chan<- string, <-chan bool) {
+	input := make(chan string)
+	done := make(chan bool)
 
-	group := sync.WaitGroup{}
+	log.Debugf("Describing ECS services")
 
-	err := c.client.ListServicesPages(
-		&ecs.ListServicesInput{Cluster: &c.cluster},
-		func(page *ecs.ListServicesOutput, lastPage bool) bool {
-			// describe each page of 10 (the max for one describe command) concurrently
-			group.Add(1)
-			serviceArns := page.ServiceArns
-			go func() {
-				defer group.Done()
-
-				resp, err := c.client.DescribeServices(&ecs.DescribeServicesInput{
-					Cluster:  &c.cluster,
-					Services: serviceArns,
-				})
-				if err != nil {
-					log.Warnf("Error describing some ECS services, ECS service report may be incomplete: %v", err)
-					return
-				}
+	go func() {
+		const maxServices = 10 // How many services we can put in one Describe command
+		group := sync.WaitGroup{}
+		lock := sync.Mutex{} // mediates access to the service cache when writing results
 
-				for _, failure := range resp.Failures {
-					log.Warnf("Failed to describe ECS service %s, ECS service report may be incomplete: %s", failure.Arn, failure.Reason)
-				}
+		describePage := func(arns []string) {
```

Comment:
To make this even easier to read I would move `describePage` out of `describeServices`. Also, instead of initializing the lock every time, wouldn't it be cleaner if it lived on the client struct?